[clang][NFC] Refactor `clang/test/SemaCXX/type-traits.cpp` to use modern `static_assert`

AMP999 opened the pull request:
static_assert is supported in all modes, so the T/F macros aren't needed anymore.

AMP999 committed:
[clang][NFC] Refactor old-style `int[__trait(T) ? 1 : -1]`  with `static_assert`

AMP999 committed:
[clang][NFC] Simply use `__has_unique_object_representations` builtin instead of `has_unique_object_representations` struct

AMP999 committed:
[clang][NFC] Use `__is_base_of` builtin in lieu of `isBaseOfT` and `isBaseOfF`

cor3ntin commented on clang/test/SemaCXX/type-traits.cpp:
```diff
@@ -2671,212 +2660,212 @@ struct has_unique_object_representations {
   static const bool value = __has_unique_object_representations(T);
 };
```
This is no longer needed at all

erichkeane replied:
The intent here was to emulate how the library would use the builtin as a type-trait.  

Also, see this bug here: https://github.com/llvm/llvm-project/issues/77585 for additional comments.

cor3ntin replied:
Except we don't do that for any other trait in that file.

erichkeane replied:
Yep, thats true too, it was the intent at the time though (and to better emulate hte test cases we were given, note that the whole section is basically a copy/paste + de-c++17'izing of what STL gave me).

My comment on the bug has more context.

cor3ntin commented on clang/test/SemaCXX/type-traits.cpp:
```diff
@@ -214,56 +212,56 @@ struct HasVirtBase : virtual ACompleteType {};
-  { int arr[F(__is_pod(AnIncompleteType))]; } // expected-error {{incomplete type}}
-  { int arr[F(__is_pod(AnIncompleteType[]))]; } // expected-error {{incomplete type}}
-  { int arr[F(__is_pod(AnIncompleteType[1]))]; } // expected-error {{incomplete type}}
+  static_assert(__is_pod(int), "");
```
Maybe we should compile with `-Wno-c++17-extensions` and get rid of the empty message

AMP999 replied:
@cor3ntin Empty messages have been eliminated by commit `5e043de`. Should we also remove the messages in the `__has_unique_object_representations` section?

cor3ntin reviewed (commented):
Looks like a nice improvement.
I have a couple of suggestions

AMP999 committed:
[Clang] Replace `has_unique_object_representations` with the builtin for abominable types

With the #81298 being approved, as soon as it gets merged into the `main` branch,
we can completely remove the `has_unique_object_representations` struct and use the builtin instead.

AMP999 committed:
[Clang] Get rid of the `static_assert`s' empty messages

cor3ntin reviewed (approved):
LGTM

AMP999 committed:
Merge branch 'main' into pr6-refactor-type-trait-tests

AMP999 committed:
Eliminate remaining T/F structs

AMP999 committed:
Merge branch 'main' into pr6-refactor-type-trait-tests

AMP999 committed:
Remove `-Wno-c++17-extensions` from `RUN` lines that are asking `C++17` or later

AMP999 commented:
@cor3ntin All checks are green now. Could you help me land this patch for me?
